File: src/main/java/com/amazonaws/kinesisvideo/signaling/tyrus/SignalingServiceWebSocketClient.java
package com.amazonaws.kinesisvideo.signaling.tyrus;

import android.util.Base64;
import android.util.Log;
import com.amazonaws.kinesisvideo.signaling.SignalingListener;

import com.amazonaws.kinesisvideo.signaling.model.Message;
import com.google.gson.Gson;
import org.glassfish.tyrus.client.ClientManager;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * Signaling service client based on websocket.
 */

public class SignalingServiceWebSocketClient {

    private static final String TAG = "SignalingServiceWebSocketClient";

    private final WebSocketClient websocketClient;

    private final ExecutorService executorService;

    private final Gson gson = new Gson();

    public SignalingServiceWebSocketClient(final String uri, final SignalingListener signalingListener,
                                           final ExecutorService executorService) {
        Log.d(TAG, "Connecting to URI " + uri + " as master");
        websocketClient = new WebSocketClient(uri, new ClientManager(), signalingListener, executorService);
        this.executorService = executorService;
    }

    public boolean isOpen() {
        return websocketClient.isOpen();
    }

    public void sendSdpOffer(final Message offer) {
        executorService.submit(new Runnable() {
            @Override
            public void run() {
                if (offer.getAction().equalsIgnoreCase("SDP_OFFER")) {

                    Log.d(TAG, "Sending Offer");

                    send(offer);
                }
            }
        });
    }

    public void sendSdpAnswer(final Message answer) {
        executorService.submit(new Runnable() {
            @Override
            public void run() {
                if (answer.getAction().equalsIgnoreCase("SDP_ANSWER")) {

                    Log.d(TAG, "Answer sent " + new String(Base64.decode(answer.getMessagePayload().getBytes(),
                            Base64.NO_WRAP | Base64.NO_PADDING | Base64.URL_SAFE)));

                    send(answer);
                }
            }
        });
    }

    public void sendIceCandidate(final Message candidate) {
        executorService.submit(new Runnable() {
            @Override
            public void run() {
                if (candidate.getAction().equalsIgnoreCase("ICE_CANDIDATE")) {

                    send(candidate);
                }

                Log.d(TAG, "Sent Ice candidate message");
            }
        });
    }

    public void disconnect() {
        executorService.submit(new Runnable() {
            @Override
            public void run() {
                websocketClient.disconnect();
            }
        });
        try {
            executorService.awaitTermination(1, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            Log.e(TAG, "Error in disconnect");
        }
    }

    private void send(final Message message) {
        String jsonMessage = gson.toJson(message);
        Log.d(TAG, "Sending JSON Message= " + jsonMessage);
        websocketClient.send(jsonMessage);
        Log.d(TAG, "Sent JSON Message= " + jsonMessage);
    }

}

File: app/src/main/java/com/amazonaws/kinesisvideo/signaling/tyrus/SignalingServiceWebSocketClient.kt
package com.amazonaws.kinesisvideo.signaling.tyrus

import android.util.Base64
import android.util.Log
import com.amazonaws.kinesisvideo.signaling.SignalingListener
import com.amazonaws.kinesisvideo.signaling.model.Message
import com.google.gson.Gson
import org.glassfish.tyrus.client.ClientManager
import java.util.concurrent.ExecutorService
import java.util.concurrent.TimeUnit

/**
 * Signaling service client based on websocket.
 */
class SignalingServiceWebSocketClient(
    uri: String,
    signalingListener: SignalingListener?,
    executorService: ExecutorService
) {
    private val websocketClient: WebSocketClient
    private val executorService: ExecutorService
    private val gson = Gson()
    private val TAG = "SignalingServiceWebSocketClient"

    init {
        Log.d(TAG, "Connecting to URI $uri as master")
        websocketClient = WebSocketClient(uri, ClientManager(), signalingListener!!, executorService)
        this.executorService = executorService
    }

    val isOpen = websocketClient.isOpen()

    fun sendSdpOffer(offer: Message) {
        executorService.submit {
            if (offer.action.equals("SDP_OFFER", ignoreCase = true)) {
                Log.d(TAG, "Sending Offer")
                send(offer)
            }
        }
    }

    fun sendSdpAnswer(answer: Message) {
        executorService.submit {
            if (answer.action.equals("SDP_ANSWER", ignoreCase = true)) {
                Log.d(
                    TAG, "Answer sent " + String(
                        Base64.decode(
                            answer.messagePayload!!.toByteArray(),
                            Base64.NO_WRAP or Base64.URL_SAFE
                        )
                    )
                )
                send(answer)
            }
        }
    }

    fun sendIceCandidate(candidate: Message) {
        executorService.submit {
            if (candidate.action.equals("ICE_CANDIDATE", ignoreCase = true)) {
                send(candidate)
            }
            Log.d(TAG, "Sent Ice candidate message")
        }
    }

    fun disconnect() {
        executorService.submit { websocketClient.disconnect() }
        try {
            executorService.shutdown()
            if (!executorService.awaitTermination(1, TimeUnit.SECONDS)) {
                executorService.shutdownNow()
            }
        } catch (e: InterruptedException) {
            Log.e(TAG, "Error in disconnect")
        }
    }

    private fun send(message: Message) {
        val jsonMessage = gson.toJson(message)
        Log.d(TAG, "Sending JSON Message= $jsonMessage")
        websocketClient.send(jsonMessage)
        Log.d(TAG, "Sent JSON Message= $jsonMessage")
    }
}